### ISSUE TITLE
More floating point analysis issues (continuing from #3191)

### ISSUE THREAD

**Issue by alexrp** [Vector35/binaryninja-api#4801]
**Version and Platform (required):**
- Binary Ninja Version: 3.6.4706-dev Personal (c4b07374)
- OS: Windows
- OS Version: 10.0.23595.1001
- CPU Architecture: x64

After #3191 getting fixed (🎉!) floating point parameter detection has improved considerably, but there's still room for improvement in some areas. I figure I'll start by listing them here and they can be broken into separate issues as/if needed. I'll be looking at [this executable](https://drive.google.com/file/d/1lYSup4dxBDhNXyY1pRroylAA6KfIeHRD/view?usp=sharing).

The function at `0x7ff69b45e7f0`:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/26f9c0af-dc61-40f0-bcce-742339d6ed1f)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/5776c11b-cc58-4192-ae97-22a122ec0539)

`arg2` should just be `float`. I can see why it's hard for BN to work this out here, because `arg2` is a fallback value that's only used with `movaps`, so on the surface it seems like the whole register is being used. That said, the fact that the function uses `movss xmm0, dword [rbx+0x4]` to set the return value could be a decent hint that the `movaps xmm0, xmm1` is only interested in the first single-precision value in `xmm1`. I don't know how hard it would be to incorporate this into BN's analysis.

The good news is that changing the return type and `arg2` to `float` makes the function read mostly as it should:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/91e10102-aa85-4332-9064-6d2b5bd7daff)

However, the operation at `0x7ff69b45e823` is not correct: It's reading a `float`, not `int32_t`, and this I suspect is a fairly straightforward issue to fix since the instruction is `movss xmm0, dword [rbx+0x4]`.

Next, let's look at `0x7ff69c0d3430`:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/12633c5a-03c2-4607-a0a8-a2fb9aac7e85)
(For brevity, I'll omit screenshots of the whole function.)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/0898a0d8-aee7-4c53-8515-8cdd63636889)

Here, we run into another issue where `movaps` gives BN the impression that the full register is being used. `arg3` and `arg4` are in fact just `float`s and are mostly used with `*ss` instructions as you'd expect. Unfortunately, they're also read with `movaps` at the beginning of the function, making BN think that they're `float[0x4]` (which is arguably kind of nonsensical anyway since C arrays are passed by reference, not value, but that's a whole separate issue).

Again, retyping the arguments to `float` (and introducing a `struct FVector { float x, y, z; }` type) results in a more correct result:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/8ae5c519-da4c-4df9-8e04-4631704bf21a)

My only (minor) complaint here would be that BN could make these SSE intrinsics more readable by mapping them to `fminf()`/`fmaxf()` (and so on). The `xorps` + `unpcklps` zero-initialization pattern also seems like something that'd be interesting to handle. (There's probably something to be said for the iteration pattern involving `rcx_2` but that's entirely unrelated to floating point stuff.)

Next, the function at `0x7ff69c0ecd90`:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/f794c534-e9f0-4ece-8c30-53340681a6d1)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/8281c200-e0b7-4502-bf46-e99a5d45864c)

BN is seemingly again confused by the `movaps` access of `xmm6` here (and later in the function), and ends up thinking that all 128 bits of `xmm6` are used, even though the function only truly works with the lower 32 `float` bits. I tried replacing that `movaps` with `nop`s to see what the decompilation would then look like:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/0c383a0d-6c04-489c-8806-803445f70640)

This is much better, though it should of course be `float` rather than `int32_t`.

Undoing the `nop` patch, retyping `arg2` to `FVector*`, and the `xmm6` variable to `float` (in that order) gets this odd result:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/a861e2ff-583e-48de-9556-ece1acc3d347)

Bizarrely, retyping `xmm6` to `float` and *then* retyping `arg2` to `FVector*` works as expected:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/6206e00b-6c9d-4204-935f-bffc0f102f8c)

Next, `0x7ff69c0ecd30`:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/d46f0b3f-1363-48ad-899b-f215e13a1931)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/ac1d9570-c523-4d34-bd38-85e4afde385e)

Here we another case where use of packed instructions confuses BN. This function is only manipulating `float` values, but the `movaps` + `xorps` leads BN to think otherwise.

Retyping stuff as needed gets a mostly good result:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/a0ec3bdc-1846-4d5f-a107-06a410bd3824)

Although we still have that odd issue where BN seems to think that `arg2->x` is being read multiple times.

That's all for now, I think. Minor issues aside, the overarching theme here seems to be that BN could (hopefully?) handle the mixing of scalar and packed instructions better.
